group-coordinator/src/test/java/org/apache/kafka/coordinator/group/GroupMetadataManagerTest.java

```diff
@@ -9607,6 +9607,151 @@ public void testOnConsumerGroupStateTransitionOnLoading() {
+        assertEquals(Collections.emptyList(), records);
+
+        // A classic group can't be downgraded.
+        context.groupMetadataManager.maybeDowngradeEmptyGroup(classicGroupId, records);
+        assertEquals(Collections.emptyList(), records);
+
+        // An empty consumer group can be upgraded.
+        context.groupMetadataManager.maybeDowngradeEmptyGroup(consumerGroupId, records);
+        assertEquals(Arrays.asList(
+            RecordHelpers.newTargetAssignmentEpochTombstoneRecord(consumerGroupId),
+            RecordHelpers.newGroupSubscriptionMetadataTombstoneRecord(consumerGroupId),
+            RecordHelpers.newGroupEpochTombstoneRecord(consumerGroupId)), records);
+        records.clear();
+
+        // A non-empty consumer group can't be downgraded.
+        ConsumerGroupMember.Builder memberBuilder = new ConsumerGroupMember.Builder(Uuid.randomUuid().toString());
+        context.replay(RecordHelpers.newMemberSubscriptionRecord(consumerGroupId, memberBuilder.build()));
+        context.groupMetadataManager.maybeDowngradeEmptyGroup(classicGroupId, records);
+        assertEquals(Collections.emptyList(), records);
+    }
+
+    @Test
+    public void testConsumerGroupHeartbeatWithEmptyClassicGroup() {
+        String classicGroupId = "classic-group-id";
+        String memberId = Uuid.randomUuid().toString();
+        MockPartitionAssignor assignor = new MockPartitionAssignor("range");
+        assignor.prepareGroupAssignment(new GroupAssignment(Collections.emptyMap()));
+        GroupMetadataManagerTestContext context = new GroupMetadataManagerTestContext.Builder()
+            .withAssignors(Collections.singletonList(assignor))
+            .build();
+        ClassicGroup classicGroup = new ClassicGroup(
+            new LogContext(),
+            classicGroupId,
+            EMPTY,
+            context.time,
+            context.metrics
+        );
+        context.replay(RecordHelpers.newGroupMetadataRecord(classicGroup, classicGroup.groupAssignment(), MetadataVersion.latestTesting()));
+
+        context.groupMetadataManager.getOrMaybeCreateClassicGroup(classicGroupId, false).transitionTo(PREPARING_REBALANCE);
+        assertThrows(GroupIdNotFoundException.class, () ->
+            context.consumerGroupHeartbeat(
+                new ConsumerGroupHeartbeatRequestData()
+                    .setGroupId(classicGroupId)
+                    .setMemberId(memberId)
+                    .setMemberEpoch(0)
+                    .setServerAssignor("range")
+                    .setRebalanceTimeoutMs(5000)
+                    .setSubscribedTopicNames(Arrays.asList("foo", "bar"))
+                    .setTopicPartitions(Collections.emptyList())));
+
+        context.groupMetadataManager.getOrMaybeCreateClassicGroup(classicGroupId, false).transitionTo(EMPTY);
+        CoordinatorResult<ConsumerGroupHeartbeatResponseData, Record> result = context.consumerGroupHeartbeat(
```

Comment:
Talking about records, I was actually wondering how their replay could work in this case. My understanding is that (1) we remove the classic group from the map directly, (2) then create the consumer group, (3) then replay the classic group deletion tombstone and (4) finally replay the consumer group records. I was expecting (3) to fail because the classic group is not there anymore. It actually works because the deletion does not check the group type in this case. I wonder if we should strengthen this path. For consumer group, the replay of tombstones only work if the group type is correct.

Reply:
Got it. Let me add a type check when replaying classic group tombstone. Now it will be like (1) remove the classic group from the map, (2) create the consumer group, (3) the classic group tombstone replay fails because the group is no longer a classic group and (4) replay the consumer group records, which only updates the group epoch because the consumer group already exists.

Reply:
Yeah, the replay will fail if you add the check. Need to think about it.

Reply:
Maybe we can remove the group only if the group is classic when replaying GroupMetadataTombstone. It's also a bit strange but I can't think of another way..

Reply:
Maybe we can remove the group only if the group is classic when replaying group metadata tombstone. It's a little strange but I can't think of another way

Reply:
I think that the weirdness comes from the fact that `getOrMaybeCreateConsumerGroup` actually put the new groups into the map. Ideally, the entire consumerGroupHeartbeat method should not alter the timeline data structure and only rely on the replay to update it. I wonder if we should fix this. If we do, they the replay in this case would be just fine. If we do this, it also means that we don't have to remove the classic group from the map too. We only need to generate the tombstone.

We don't have this issue on the classic group case because we don't replay records.